### PR TITLE
U2: Port SpinnerView and ScrollView to U2.

### PR DESCRIPTION
**Pull request by shepheb**

### CODE REVIEW

File: js/foam/u2/ScrollView.js [hunk 1]
```diff
@@ -0,0 +1,617 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (kgrgreer) [inline js/foam/u2/ScrollView.js]: Year is 2016.

File: js/foam/u2/ScrollView.js [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+      postSet: function(_, nu) {
+        if (typeof nu === 'string' && this.rowHeight < 0) {
+          var view = this.Y.lookup(nu);
+          if ( view.PREFERRED_HEIGHT )
+            this.rowHeight = view.create({ model: this.data.model }).preferredHeight;
+        }
+      }
+    },
+    {
+      name: '$container',
```
Comment (kgrgreer) [inline js/foam/u2/ScrollView.js]: This naming made sense in u1 when we accessed elements with .$, but in U2 this doesn't exist, so a better convention would be to name this containerE, scrollerE, etc.

File: js/foam/u2/ScrollView.js [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+      factory: function() { return {}; }
+    },
+    {
+      name: 'extraRows',
+      documentation: 'Buffer of extra, unneeded visible rows. These will be the first to be reused if more rows are needed.',
+      factory: function() { return []; }
+    },
+    {
+      name: 'cache',
+      model_: 'ArrayProperty',
```
Comment (kgrgreer) [inline js/foam/u2/ScrollView.js]: type: 'Array'

File: js/foam/u2/ScrollView.js [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+        this.visibleTop = Math.max(0, this.visibleIndex - runwayCount);
+        this.visibleBottom = Math.min(this.count - 1,
+            this.visibleIndex + Math.ceil( (this.runway + this.viewportHeight) / this.rowHeight ) );
+
+        // Now, if the visible range is truncated, expand it. The only cases where truncation
+        // can happen is if we're abutting one edge of the range or the other, so we just extend
+        // the opposite end of the range until it fits the maximum set of rows.
+        var maxVisible = Math.ceil((2 * this.runway + this.viewportHeight) / this.rowHeight);
+        if ( this.visibleBottom - this.visibleTop + 1 < maxVisible ) {
+          if ( this.visibleTop === 0 ) this.visibleBottom = Math.min(maxVisible - 1, this.count - 1);
```
Comment (kgrgreer) [inline js/foam/u2/ScrollView.js]: Format this if/else on 4 lines.  This formatting is hard to read.

File: js/foam/u2/SpinnerView.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (kgrgreer) [inline js/foam/u2/SpinnerView.js]: 2016

**Comment by shepheb** [comment]
All addressed, PTAL.

**Pull request merged**